File: plugins/wise/engine/src/channel.ts
// Child channel helpers (research-ts-engine.md P8, D16): live status derived from a child's event
// stream, the throttle behind `step.progress`, the stale watchdog, report clipping, the context
// lookup behind `wise_context`, checkpoints, and the synchronous-mode answer for `wise_ask`.
// Everything here is pure or injectable; the executor wires it to real children and the ledger.

import { mkdirSync, renameSync, rmSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { headline } from "./steps/agent.ts";
import type { ChildProgress, RawEvent, State } from "./types.ts";

export const PROGRESS_THROTTLE_MS = 30_000;
export const STALE_AFTER_SECS_DEFAULT = 600;
export const REPORT_TEXT_MAX = 200;
export const REPORT_DATA_MAX_BYTES = 1024;

type Rec = Record<string, unknown>;
const isRec = (v: unknown): v is Rec => typeof v === "object" && v !== null && !Array.isArray(v);
const num = (v: unknown): number => (typeof v === "number" && Number.isFinite(v) ? v : 0);

// ---- timers ---------------------------------------------------------------------------------------

/** Clock and timers, injectable so tests drive the stale policy without waiting. */
export type ChannelTimers = {
  now: () => number;
  setTimeout: (fn: () => void, ms: number) => unknown;
  clearTimeout: (handle: unknown) => void;
};

export const realTimers: ChannelTimers = {
  now: () => Date.now(),
  setTimeout: (fn, ms) => {
    const t = setTimeout(fn, ms);
    t.unref?.();
    return t;
  },
  clearTimeout: (h) => clearTimeout(h as NodeJS.Timeout),
};

// ---- live status from the stream --------------------------------------------------------------------

/** Sum of the `*_tokens` counters in a vendor usage object (Claude and Codex shapes). */
function usageTokens(u: unknown): number {
  if (!isRec(u)) return 0;
  let total = 0;
  for (const [k, v] of Object.entries(u)) if (k.endsWith("_tokens")) total += num(v);
  return total;
}

const DETAIL_MAX = 80;
const TEXT_MAX = 100;

/** Input keys that say what a tool touched, in preference order (Claude Code tool shapes). */
const DETAIL_KEYS = [
  "file_path",
  "path",
  "notebook_path",
  "command",
  "pattern",
  "query",
  "url",
  "skill",
  "description",
  "prompt",
];

function firstLine(text: string, max: number): string {
  const line = text.replace(/\s+/g, " ").trim();
  return line.length > max ? `${line.slice(0, max - 1)}…` : line;
}

/** `{name, detail?}` of the last `tool_use` block, `detail` = the most telling input value. */
function lastToolUse(message: unknown): { name: string; detail?: string } | undefined {
  if (!isRec(message) || !Array.isArray(message.content)) return undefined;
  const uses = message.content.filter((b): b is Rec => isRec(b) && b.type === "tool_use");
  const last = uses[uses.length - 1];
  if (!last) return undefined;
  const out: { name: string; detail?: string } = {
    name: typeof last.name === "string" ? last.name : "?",
  };
  if (isRec(last.input)) {
    for (const key of DETAIL_KEYS) {
      const v = last.input[key];
      if (typeof v === "string" && v.trim()) {
        out.detail = firstLine(v, DETAIL_MAX);
        break;
      }
    }
  }
  return out;
}

/** Headline of the last `text` block of an assistant message, or `undefined`. */
function lastText(message: unknown): string | undefined {
  if (!isRec(message) || !Array.isArray(message.content)) return undefined;
  const texts = message.content.filter(
    (b): b is Rec => isRec(b) && b.type === "text" && typeof b.text === "string",
  );
  const last = texts[texts.length - 1];
  if (!last) return undefined;
  const line = firstLine(last.text as string, TEXT_MAX);
  return line || undefined;
}

export type ChildTracker = {
  /** Feed one vendor event; returns a snapshot when a `step.progress` event is due. */
  ingest: (e: RawEvent) => ChildProgress | undefined;
  /** A `wise_report` arrived. */
  report: () => void;
  /** Any other channel call: counts as activity. */
  touch: () => void;
  snapshot: () => ChildProgress;
  lastActivityMs: () => number;
};

export type TrackerOpts = {
  step: string;
  throttleMs?: number;
  /** Minimum gap between two emits triggered by a detail (not tool) change; default 5 000. */
  detailGapMs?: number;
  now?: () => number;
};

export const DETAIL_GAP_MS = 5_000;

/**
 * Derives turn count, last tool and what it touched, latest assistant text, tokens so far and
 * last activity from the raw stream. Emits a progress snapshot at once when the tool name
 * changes, when the tool's target changes (at most one per `detailGapMs`), and otherwise at most
 * once per `throttleMs` per step.
 */
export function createChildTracker(opts: TrackerOpts): ChildTracker {
  const now = opts.now ?? (() => Date.now());
  const throttleMs = opts.throttleMs ?? PROGRESS_THROTTLE_MS;
  const detailGapMs = opts.detailGapMs ?? DETAIL_GAP_MS;
  let turn = 0;
  let tool: string | undefined;
  let detail: string | undefined;
  let text: string | undefined;
  let tokens = 0;
  let reports = 0;
  const started = now();
  let lastActivity = started;
  let lastEmit = lastActivity;

  const snapshot = (): ChildProgress => {
    const snap: ChildProgress = {
      step: opts.step,
      turn,
      tokens,
      last_activity: new Date(lastActivity).toISOString(),
      reports,
      elapsed_ms: lastActivity - started,
    };
    if (tool !== undefined) snap.tool = tool;
    if (detail !== undefined) snap.detail = detail;
    if (text !== undefined) snap.text = text;
    return snap;
  };

  return {
    ingest(e) {
      lastActivity = now();
      const parsed = e.parsed;
      if (!isRec(parsed)) return undefined;
      const beforeTool = tool;
      const beforeDetail = detail;
      if (parsed.type === "assistant") {
        turn += 1;
        const use = lastToolUse(parsed.message);
        if (use) {
          tool = use.name;
          detail = use.detail;
        }
        const t = lastText(parsed.message);
        if (t !== undefined) text = t;
        if (isRec(parsed.message)) tokens += usageTokens(parsed.message.usage);
      } else if (parsed.type === "result") {
        // Cumulative totals from the harness win over the per-turn sum.
        tokens = Math.max(tokens, usageTokens(parsed.usage));
      } else if (parsed.usage !== undefined) {
        tokens += usageTokens(parsed.usage);
      }
      const since = lastActivity - lastEmit;
      const toolChanged = tool !== beforeTool;
      const detailChanged = !toolChanged && detail !== beforeDetail && since >= detailGapMs;
      if (!toolChanged && !detailChanged && since < throttleMs) return undefined;
      lastEmit = lastActivity;
      return snapshot();
    },
    report() {
      reports += 1;
      lastActivity = now();
    },
    touch() {
      lastActivity = now();
    },
    snapshot,
    lastActivityMs: () => lastActivity,
  };
}

function fmtTokens(n: number): string {
  return n >= 1000 ? `${Math.round(n / 1000)}k` : String(n);
}

function fmtElapsed(ms: number): string {
  const s = Math.floor(ms / 1000);
  if (s < 60) return `${s}s`;
  const m = Math.floor(s / 60);
  if (m < 60) return `${m}m${String(s % 60).padStart(2, "0")}s`;
  return `${Math.floor(m / 60)}h${String(m % 60).padStart(2, "0")}m`;
}

/**
 * One-line rendering for the `step.progress` event:
 * `turn 14, tool Edit src/x.ts, 48k tokens, 1 report, 2m10s: <latest assistant text>`.
 */
export function progressLine(p: ChildProgress): string {
  const parts = [`turn ${p.turn}`];
  if (p.tool !== undefined) parts.push(`tool ${p.tool}${p.detail ? ` ${p.detail}` : ""}`);
  parts.push(`${fmtTokens(p.tokens)} tokens`);
  if (p.reports > 0) parts.push(`${p.reports} report${p.reports === 1 ? "" : "s"}`);
  if (p.elapsed_ms !== undefined) parts.push(fmtElapsed(p.elapsed_ms));
  const line = parts.join(", ");
  return p.text ? `${line}: ${p.text}` : line;
}

// ---- stale watchdog ------------------------------------------------------------------------------------

export type StaleWatchOpts = {
  staleMs: number;
  timers: ChannelTimers;
  lastActivityMs: () => number;
  /** Idle because a human has not answered its question yet: never stale. */
  paused?: () => boolean;
  /** Deliver the idle nudge; `false` when the harness has no stdin (kill at once, E8). */
  nudge: (idleMs: number) => boolean;
  kill: () => void;
};

export type StaleWatch = { stop: () => void };

/**
 * After `staleMs` of no activity: one nudge when the harness takes one, then a kill after another
 * idle window. Activity between the two resets the policy, so each idle stretch gets its own nudge.
 */
export function startStaleWatch(opts: StaleWatchOpts): StaleWatch {
  const { timers, staleMs } = opts;
  let handle: unknown;
  let nudged = false;
  let stopped = false;

  const arm = (ms: number): void => {
    if (stopped) return;
    handle = timers.setTimeout(fire, Math.max(1, ms));
  };
  const fire = (): void => {
    if (stopped) return;
    if (opts.paused?.()) {
      arm(staleMs);
      return;
    }
    const idle = timers.now() - opts.lastActivityMs();
    if (idle < staleMs) {
      nudged = false;
      arm(staleMs - idle);
      return;
    }
    if (!nudged && opts.nudge(idle)) {
      nudged = true;
      arm(staleMs);
      return;
    }
    stopped = true;
    opts.kill();
  };
  arm(staleMs);
  return {
    stop() {
      stopped = true;
      if (handle !== undefined) timers.clearTimeout(handle);
    },
  };
}

export function staleNudgeText(idleMs: number): string {
  const minutes = Math.max(1, Math.round(idleMs / 60_000));
  return `You have been idle for ${minutes} minute${minutes === 1 ? "" : "s"}. Finish with your structured result now.`;
}

// ---- reports ---------------------------------------------------------------------------------------------

export function clipReportText(text: string): string {
  return headline(text, REPORT_TEXT_MAX);
}

/** The report payload when it serializes under the cap, otherwise nothing. */
export function clipReportData(data: unknown): Record<string, unknown> | undefined {
  if (!isRec(data)) return undefined;
  let json: string;
  try {
    json = JSON.stringify(data);
  } catch {
    return undefined;
  }
  return Buffer.byteLength(json, "utf8") <= REPORT_DATA_MAX_BYTES ? data : undefined;
}

// ---- context lookup --------------------------------------------------------------------------------------

function pathLookup(root: unknown, path: string[]): unknown {
  let cur: unknown = root;
  for (const seg of path) {
    if (Array.isArray(cur) && /^\d+$/.test(seg)) cur = cur[Number(seg)];
    else if (isRec(cur)) cur = cur[seg];
    else return undefined;
    if (cur === undefined) return undefined;
  }
  return cur;
}

/**
 * `wise_context(key)`: run context first (`ticket`, `guidance`, `decisions`, `links`, dotted paths
 * such as `ticket.0.body`), then run outputs, a step's outputs by step id, then inputs. `null` when
 * nothing matches.
 */
export function resolveContextKey(state: State, key: string): unknown {
  const path = key.split(".").filter((s) => s.length > 0);
  if (path.length === 0) return null;
  const fromContext = pathLookup(state.context, path);
  if (fromContext !== undefined) return fromContext;
  const fromOutputs = pathLookup(state.outputs, path);
  if (fromOutputs !== undefined) return fromOutputs;
  const [stepId, ...rest] = path;
  const step = stepId !== undefined ? state.steps[stepId] : undefined;
  if (step?.outputs !== undefined) {
    const fromStep = rest.length === 0 ? step.outputs : pathLookup(step.outputs, rest);
    if (fromStep !== undefined) return fromStep;
  }
  const fromInputs = pathLookup(state.inputs, path);
  return fromInputs === undefined ? null : fromInputs;
}

// ---- checkpoints ------------------------------------------------------------------------------------------

export function checkpointPath(runDir: string, stepId: string): string {
  return join(runDir, "checkpoints", `${stepId}.json`);
}

/** Atomic write of the child's partial result; survives a stale kill. Returns the file path. */
export function writeCheckpoint(runDir: string, stepId: string, data: unknown): string {
  const path = checkpointPath(runDir, stepId);
  mkdirSync(join(runDir, "checkpoints"), { recursive: true });
  const tmp = `${path}.tmp`;
  try {
    writeFileSync(tmp, JSON.stringify(data ?? null, null, 2) + "\n", "utf8");
    renameSync(tmp, path);
  } catch (err) {
    rmSync(tmp, { force: true });
    throw err;
  }
  return path;
}

// ---- synchronous answers -----------------------------------------------------------------------------------

const norm = (s: string): string => s.trim().toLowerCase();

/**
 * Best-effort answer for `wise_ask` in a synchronous run: a decision keyed by the question text,
 * then a decision whose key appears in the question (or the reverse), then a decision whose value
 * is one of the options, then the first option. `undefined` means a human is needed.
 */
export function answerFromDecisions(
  question: string,
  options: readonly string[] | undefined,
  decisions: Record<string, string> | undefined,
): string | undefined {
  const q = norm(question);
  const entries = Object.entries(decisions ?? {});
  const exact = entries.find(([k]) => norm(k) === q);
  if (exact) return exact[1];
  const partial = entries.find(([k]) => {
    const nk = norm(k);
    return nk.length > 0 && (q.includes(nk) || nk.includes(q));
  });
  if (partial) return partial[1];
  if (options && options.length > 0) {
    const byOption = entries.find(([, v]) => options.some((o) => norm(o) === norm(v)));
    if (byOption) return byOption[1];
    return options[0];
  }
  return undefined;
}

File: plugins/wise/engine/wise_engine/channel.py
from __future__ import annotations

import asyncio
import math
import re
import time
from collections.abc import Callable
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from .ledger import write_json_atomic
from .render import _json
from .scheduler import JS_WHITESPACE, UNDEFINED, number_text
from .steps.agent import headline, utf16_length, utf16_slice

PROGRESS_THROTTLE_MS = 30_000
STALE_AFTER_SECS_DEFAULT = 600
REPORT_TEXT_MAX = 200
REPORT_DATA_MAX_BYTES = 1024
DETAIL_MAX = 80
TEXT_MAX = 100
DETAIL_GAP_MS = 5_000
DETAIL_KEYS = (
    "file_path",
    "path",
    "notebook_path",
    "command",
    "pattern",
    "query",
    "url",
    "skill",
    "description",
    "prompt",
)
_SPACE = re.compile(f"[{re.escape(JS_WHITESPACE)}]+")
Json = dict[str, Any]


def now_ms() -> float:
    return time.time() * 1000


@dataclass
class ChannelTimers:
    now: Callable[[], float]
    set_timeout: Callable[[Callable[[], None], float], Any]
    clear_timeout: Callable[[Any], None]


def real_timers() -> ChannelTimers:
    return ChannelTimers(
        now_ms,
        lambda fn, ms: asyncio.get_running_loop().call_later(ms / 1000, fn),
        lambda handle: handle.cancel(),
    )


def _number(value: Any) -> float:
    return value if type(value) in (int, float) and math.isfinite(value) else 0


def _usage_tokens(usage: Any) -> float:
    return (
        sum(_number(value) for key, value in usage.items() if key.endswith("_tokens"))
        if isinstance(usage, dict)
        else 0
    )


def _first_line(text: str, maximum: int) -> str:
    line = _SPACE.sub(" ", text).strip(JS_WHITESPACE)
    return utf16_slice(line, 0, maximum - 1) + "…" if utf16_length(line) > maximum else line


def _last_tool_use(message: Any) -> Json | None:
    if not isinstance(message, dict) or not isinstance(message.get("content"), list):
        return None
    uses = [b for b in message["content"] if isinstance(b, dict) and b.get("type") == "tool_use"]
    if not uses:
        return None
    last = uses[-1]
    result = {"name": last["name"] if isinstance(last.get("name"), str) else "?"}
    if isinstance(last.get("input"), dict):
        for key in DETAIL_KEYS:
            value = last["input"].get(key)
            if isinstance(value, str) and value.strip(JS_WHITESPACE):
                result["detail"] = _first_line(value, DETAIL_MAX)
                break
    return result


def _last_text(message: Any) -> str | None:
    if not isinstance(message, dict) or not isinstance(message.get("content"), list):
        return None
    texts = [
        b["text"]
        for b in message["content"]
        if isinstance(b, dict) and b.get("type") == "text" and isinstance(b.get("text"), str)
    ]
    return (_first_line(texts[-1], TEXT_MAX) or None) if texts else None


class ChildTracker:
    def __init__(
        self,
        *,
        step: str,
        throttle_ms: float = PROGRESS_THROTTLE_MS,
        detail_gap_ms: float = DETAIL_GAP_MS,
        now: Callable[[], float] = now_ms,
    ):
        self.step = step
        self.throttle_ms = throttle_ms
        self.detail_gap_ms = detail_gap_ms
        self.now = now
        self.turn = 0
        self.tool: str | None = None
        self.detail: str | None = None
        self.text: str | None = None
        self.tokens: float = 0
        self.reports = 0
        self.started = self.last_activity = self.last_emit = now()

    def snapshot(self) -> Json:
        snap = dict(
            step=self.step,
            turn=self.turn,
            tokens=self.tokens,
            last_activity=datetime.fromtimestamp(self.last_activity / 1000, timezone.utc)
            .isoformat(timespec="milliseconds")
            .replace("+00:00", "Z"),
            reports=self.reports,
            elapsed_ms=self.last_activity - self.started,
        )
        for key in ("tool", "detail", "text"):
            value = getattr(self, key)
            if value is not None:
                snap[key] = value
        return snap

    def ingest(self, event: Json) -> Json | None:
        self.last_activity = self.now()
        parsed = event.get("parsed")
        if not isinstance(parsed, dict):
            return None
        before_tool, before_detail = self.tool, self.detail
        if parsed.get("type") == "assistant":
            self.turn += 1
            use = _last_tool_use(parsed.get("message"))
            if use:
                self.tool = use["name"]
                self.detail = use.get("detail")
            text = _last_text(parsed.get("message"))
            if text is not None:
                self.text = text
            if isinstance(parsed.get("message"), dict):
                self.tokens += _usage_tokens(parsed["message"].get("usage"))
        elif parsed.get("type") == "result":
            self.tokens = max(self.tokens, _usage_tokens(parsed.get("usage")))
        elif "usage" in parsed:
            self.tokens += _usage_tokens(parsed["usage"])
        since = self.last_activity - self.last_emit
        changed = self.tool != before_tool
        detail_changed = (
            not changed and self.detail != before_detail and since >= self.detail_gap_ms
        )
        if not changed and not detail_changed and since < self.throttle_ms:
            return None
        self.last_emit = self.last_activity
        return self.snapshot()

    def report(self) -> None:
        self.reports += 1
        self.touch()

    def touch(self) -> None:
        self.last_activity = self.now()

    def last_activity_ms(self) -> float:
        return self.last_activity


def create_child_tracker(opts: Json) -> ChildTracker:
    return ChildTracker(**opts)


def _fmt_tokens(count: float) -> str:
    return f"{math.floor(count / 1000 + 0.5)}k" if count >= 1000 else number_text(count)


def _fmt_elapsed(milliseconds: float) -> str:
    seconds = math.floor(milliseconds / 1000)
    if seconds < 60:
        return f"{seconds}s"
    minutes = seconds // 60
    return (
        f"{minutes}m{seconds % 60:02}s" if minutes < 60 else f"{minutes // 60}h{minutes % 60:02}m"
    )


def progress_line(progress: Json) -> str:
    parts = [f"turn {progress['turn']}"]
    if "tool" in progress:
        parts.append(
            f"tool {progress['tool']}"
            + (f" {progress['detail']}" if progress.get("detail") else "")
        )
    parts.append(f"{_fmt_tokens(progress['tokens'])} tokens")
    if progress["reports"] > 0:
        parts.append(f"{progress['reports']} report" + ("" if progress["reports"] == 1 else "s"))
    if "elapsed_ms" in progress:
        parts.append(_fmt_elapsed(progress["elapsed_ms"]))
    line = ", ".join(parts)
    return f"{line}: {progress['text']}" if progress.get("text") else line


class StaleWatch:
    def __init__(
        self,
        *,
        stale_ms: float,
        timers: ChannelTimers,
        last_activity_ms: Callable[[], float],
        nudge: Callable[[float], bool],
        kill: Callable[[], None],
        paused: Callable[[], bool] | None = None,
    ):
        self.stale_ms = stale_ms
        self.timers = timers
        self.last_activity_ms = last_activity_ms
        self.nudge = nudge
        self.kill = kill
        self.paused = paused
        self.nudged = False
        self.stopped = False
        self.handle = None
        self._arm(stale_ms)

    def _arm(self, ms: float) -> None:
        if not self.stopped:
            self.handle = self.timers.set_timeout(self._fire, max(1, ms))

    def _fire(self) -> None:
        if self.stopped:
            return
        if self.paused and self.paused():
            self._arm(self.stale_ms)
            return
        idle = self.timers.now() - self.last_activity_ms()
        if idle < self.stale_ms:
            self.nudged = False
            self._arm(self.stale_ms - idle)
        elif not self.nudged and self.nudge(idle):
            self.nudged = True
            self._arm(self.stale_ms)
        else:
            self.stopped = True
            self.kill()

    def stop(self) -> None:
        self.stopped = True
        if self.handle is not None:
            self.timers.clear_timeout(self.handle)


def start_stale_watch(opts: Json) -> StaleWatch:
    return StaleWatch(**opts)


def stale_nudge_text(idle_ms: float) -> str:
    minutes = max(1, math.floor(idle_ms / 60000 + 0.5))
    return (
        f"You have been idle for {minutes} minute"
        + ("" if minutes == 1 else "s")
        + ". Finish with your structured result now."
    )


def clip_report_text(text: str) -> str:
    return headline(text, REPORT_TEXT_MAX)


def clip_report_data(data: Any) -> Json | None:
    if not isinstance(data, dict):
        return None
    try:
        text = _json(data)
        return data if len(text.encode("utf-8")) <= REPORT_DATA_MAX_BYTES else None
    except (ValueError, TypeError, RecursionError, UnicodeError):
        return None


def _path_lookup(root: Any, path: list[str]) -> Any:
    current = root
    for segment in path:
        if isinstance(current, list) and re.fullmatch("[0-9]+", segment):
            index = int(segment)
            current = current[index] if index < len(current) else UNDEFINED
        elif isinstance(current, dict):
            current = current.get(segment, UNDEFINED)
        else:
            return UNDEFINED
        if current is UNDEFINED:
            return UNDEFINED
    return current


def resolve_context_key(state: Json, key: str) -> Any:
    path = [p for p in key.split(".") if p]
    if not path:
        return None
    for owner in ("context", "outputs"):
        value = _path_lookup(state.get(owner, {}), path)
        if value is not UNDEFINED:
            return value
    step = state.get("steps", {}).get(path[0])
    if isinstance(step, dict) and "outputs" in step:
        value = _path_lookup(step["outputs"], path[1:])
        if value is not UNDEFINED:
            return value
    value = _path_lookup(state.get("inputs", {}), path)
    return None if value is UNDEFINED else value


def checkpoint_path(run_dir: str | Path, step_id: str) -> str:
    return str(Path(run_dir) / "checkpoints" / f"{step_id}.json")


def write_checkpoint(run_dir: str | Path, step_id: str, data: Any) -> str:
    path = checkpoint_path(run_dir, step_id)
    write_json_atomic(path, None if data is UNDEFINED else data)
    return path


def answer_from_decisions(
    question: str, options: list[str] | None = None, decisions: dict[str, str] | None = None
) -> str | None:
    def norm(text):
        return text.strip(JS_WHITESPACE).lower()

    q = norm(question)
    entries = list((decisions or {}).items())
    for key, value in entries:
        if norm(key) == q:
            return value
    for key, value in entries:
        normalized = norm(key)
        if normalized and (normalized in q or q in normalized):
            return value
    if options:
        for _, value in entries:
            if any(norm(option) == norm(value) for option in options):
                return value
        return options[0]
    return None
